apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/service/blockchain/TransactionServiceImpl.java

```diff
@@ -0,0 +1,280 @@
+                from = 0; // set to zero for shard db
+            } else {
+                from -= foundCount;
+                to -= foundCount;
+            }
+            // loop over all shard in descent order and fetch left Tx number
+            Iterator<TransactionalDataSource> fullDataSources = ((ShardManagement) databaseManager).getAllFullDataSourcesIterator();
+            while (fullDataSources.hasNext()) {
+                TransactionalDataSource dataSource = fullDataSources.next();
+                // count Tx records before fetch Tx records
+                foundCount = transactionDao.getTransactionCountByFilter(dataSource,
+                    accountId, numberOfConfirmations, type, subtype,
+                    blockTimestamp, withMessage, phasedOnly, nonPhasedOnly,
+                    includeExpiredPrunable, executedOnly, includePrivate, height, prunableExpiration);
+                log.trace("countTx 3. DS={}, from={} to={}, foundCount={} (skip='{}')\naccountId={}, type={}, subtype={}",
+                    dataSource.getDbIdentity(), from, to, foundCount, (foundCount <= 0),
+                    accountId, type, subtype);
+                if (foundCount == 0) {
+                    continue; // skip shard without any suitable records
+                }
+                // because count is > 0 then try to fetch Tx records from shard db
+                List<TransactionEntity> fetchedTxs = transactionDao.getTransactions(
+                    dataSource,
+                    accountId, numberOfConfirmations, type, subtype,
+                    blockTimestamp, withMessage, phasedOnly, nonPhasedOnly,
+                    from, to, includeExpiredPrunable, executedOnly, includePrivate, height, prunableExpiration);
+                log.trace("getTx() 4. DS={} fetched [{}] (foundCount={}) from={}, to={}", dataSource.getDbIdentity(),
+                    fetchedTxs.size(), foundCount, from, to);
+                if (fetchedTxs.isEmpty()) {
+                    to -= foundCount;
+                    from -= foundCount;
+                } else {
+                    to -= (fetchedTxs.size() + from);
+                    from = 0;
+                }
+                transactions.addAll(fetchedTxs);
+                if (transactions.size() >= limit) { // by default, size of fetched transactions should be equal to initialLimit, but when error occurred this check allow to avoid fetching all txs
+                    break;
+                }
+            }
+        }
+        log.trace("Tx number Requested / Loaded : [{}] / [{}] = in {} ms", limit, transactions.size(), System.currentTimeMillis() - start);
+        return transactions.stream().map(toModelConverter).collect(Collectors.toList());
+    }
+
+    private TransactionalDataSource getDatasourceWithShardingByTransactionId(long transactionId) {
+        Long shardId = transactionIndexDao.getShardIdByTransactionId(transactionId);
```

Comment:
What if 'shardId' is NULL here ?

Reply:
The data source to the main database will be returned.